vector-stores/spring-ai-gemfire/src/main/java/org/springframework/ai/vectorstore/GemFireVectorStore.java

```diff
@@ -0,0 +1,531 @@
+
+		public QueryRequest(List<Float> vector, int k, int kPerBucket, boolean includeMetadata) {
+			this.vector = vector;
+			this.k = k;
+			this.kPerBucket = kPerBucket;
+			this.includeMetadata = includeMetadata;
+		}
+
+		public List<Float> getVector() {
+			return vector;
+		}
+
+		public int getK() {
+			return k;
+		}
+
+		public int getkPerBucket() {
+			return kPerBucket;
+		}
+
+		public boolean isIncludeMetadata() {
+			return includeMetadata;
+		}
+
+	}
+
+	private static final class QueryResponse {
+
+		private String key;
+
+		private float score;
+
+		private Map<String, Object> metadata;
+
+		private String getContent(String field) {
+			return (String) metadata.get(field);
+		}
+
+		public void setKey(String key) {
+			this.key = key;
+		}
+
+		public void setScore(float score) {
+			this.score = score;
+		}
+
+		public void setMetadata(Map<String, Object> metadata) {
+			this.metadata = metadata;
+		}
+
+	}
+
+	private static class DeleteRequest {
+
+		@JsonProperty("delete-data")
+		private boolean deleteData = true;
+
+		public DeleteRequest() {
+		}
+
+		public DeleteRequest(boolean deleteData) {
+			this.deleteData = deleteData;
+		}
+
+		public boolean isDeleteData() {
+			return deleteData;
+		}
+
+		public void setDeleteData(boolean deleteData) {
+			this.deleteData = deleteData;
+		}
+
+	}
+
+	@Override
+	public void add(List<Document> documents) {
+		UploadRequest upload = new UploadRequest(documents.stream().map(document -> {
+			// Compute and assign an embedding to the document.
+			document.setEmbedding(this.embeddingClient.embed(document));
+			List<Float> floatVector = document.getEmbedding().stream().map(Double::floatValue).toList();
+			return new UploadRequest.Embedding(document.getId(), floatVector, documentField, document.getContent(),
+					document.getMetadata());
+		}).toList());
+
+		ObjectMapper objectMapper = new ObjectMapper();
+		String embeddingsJson = null;
+		try {
+			String embeddingString = objectMapper.writeValueAsString(upload);
+			embeddingsJson = embeddingString.substring("{\"embeddings\":".length());
+		}
+		catch (JsonProcessingException e) {
+			throw new RuntimeException("Embedding json parsing error :" + e.getMessage());
+		}
+
+		client.post()
+			.uri("/" + indexName + EMBEDDINGS)
+			.contentType(MediaType.APPLICATION_JSON)
+			.bodyValue(embeddingsJson)
+			.retrieve()
+			.bodyToMono(Void.class)
+			.onErrorMap(WebClientException.class, this::handleHttpClientException)
+			.block();
+	}
+
+	@Override
+	public Optional<Boolean> delete(List<String> idList) {
+		try {
+			client.method(HttpMethod.DELETE)
+				.uri("/" + indexName + EMBEDDINGS)
+				.body(BodyInserters.fromValue(idList))
+				.retrieve()
+				.bodyToMono(Void.class)
+				.block();
+		}
+		catch (Exception e) {
+			logger.warn("Error removing embedding: " + e);
+			return Optional.of(false);
+		}
+		return Optional.of(true);
+	}
+
+	@Override
+	public List<Document> similaritySearch(SearchRequest request) {
```

Comment:
The SearchRequest allows metadata filtering using [filter expressions](https://docs.spring.io/spring-ai/reference/api/vectordbs.html#metadata-filters) (DSL and SQL-like text). 
All Spring AI Vector Store implementations so far have opted to provide support for this feature. 
I remember that Gemfire can have a powerful predicate filtering support so It seems possible to map the Spring AI Filter expressions to the existin Gemifre one. (Just guessing). 
But if you decide not to support it, then here you have to check if `SearchRequest.hasFilterExpression()` is used and throw an exception making it explicit that Gemfire does not support metadata filter expressions.

Reply:
We do expect to provide prefiltering based on Lucene indexes in the future. Will be nice, just iterating towards that at the moment.